keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1480 @@
+
+### API Specification
+
+#### VolumeSnapshotDelta Resource
+
+The section describes the specification of `VolumeSnapshotDelta` resource:
+
+```go
+// VolumeSnapshotDelta represents a VolumeSnapshotDelta resource.
+type VolumeSnapshotDelta struct {
+  metav1.TypeMeta `json:",inline"`
+
+  // Standard object's metadata.
+  // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // spec defines the desired characteristics of a snapshot delta requested by a user.
+  // Required.
+  Spec VolumeSnapshotDeltaSpec `json:"spec"`
+
+  // status represents the current information of a snapshot delta.
+  // +optional
+  Status VolumeSnapshotDeltaStatus `json:"status,omitempty"`
+}
+
+// VolumeSnapshotDeltaSpec is the spec of a VolumeSnapshotDelta resource.
+type VolumeSnapshotDeltaSpec struct {
+  // The name of the base CSI volume snapshot to use for comparison.
+  // If not specified, return the delta between the first snapshot
+  // since volume creation and the target snapshot.
+  // +optional
+  FromVolumeSnapshotName string `json:"fromVolumeSnapshotName,omitempty"`
+
+  // The name of the target CSI volume snapshot to use for comparison.
+  // Required.
+  ToVolumeSnapshotName string `json:"toVolumeSnapshotName"`
+
+  // VolumeSnapshotDeltaRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver to complete the CSI
+  // calls for VolumeSnapshotDelta.
+  // This field is optional, and may be empty if no secret is required. If the
+  // secret object contains more than one secret, all secrets are passed.
+  // +optional
+  VolumeSnapshotDeltaSecretRef *SecretReference
+
+  // Define the maximum number of entries to return in the response.
+  Limit uint64 `json:"limit"`
+
+  // Defines the start of the block index (in bytes).
+  OffsetBytes uint64 `json:"offsetBytes"`
+}
+
+// VolumeSnapshotDeltaStatus is the status for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaStatus struct {
+  // The list of CBT entries.
+  ChangedBlocks []*ChangedBlocks `json:"changedBlocks;omitempty"`
+
+  // Captures any error encountered.
+  Error string `json:"error,omitempty"`
+
+  // A very brief description to communicate the current state of the CBT
+  // operation.
+  State VolumeSnapshotDeltaState `json:"state"`
+
+  // The limit defined in the request.
+  Limit uint64 `json:"limit"`
+
+  // The offset (in bytes) defined in the request.
+  OffsetBytes uint64 `json:"offsetBytes"`
+
+  // The starting block index of the next request.
+  Continue uint64 `json:"continue"`
+}
+
+// ChangedBlock represents a CBT entry returned by the storage provider.
+type ChangedBlock struct {
+  // OffsetBytes defines the start of the block index in the response.
+  OffsetBytes uint64 `json:"offsetBytes"`
+
+  // The size of the blocks.
+  BlockSizeBytes unit64 `json:"blockSizeBytes"`
```

Comment:
storage is the only place where I actually worry that uint64 might eventually be too small - am I crazy?